ext/ddtrace_profiling_native_extension/stack_recorder.c

```diff
@@ -8,16 +10,154 @@
+//     `slot_two_mutex`, that one was taken. Since the serializer thread is expected only to work once a minute,
+//     we retry steps 1. and 2. and should be able to find an active slot.
+//   b. Something is incorrect in the StackRecorder state. In this situation, the sampler thread should give up on
+//     sampling and enter an error state.
+//
+// Note that in the steps above, and because the sampler thread uses `trylock` to probe the mutexes, that the
+// sampler thread never blocks. It either is able to find an active profile slot in a bounded amount of steps or it
+// enters an error state.
+//
+// This guarantees that sampler performance is never constrained by serializer performance.
+//
+// ### Locking protocol, from the serializer thread side
+//
+// When the serializer thread wants to serialize a profile, it first flips the active and inactive profile slots.
+//
+// The flipping action is described below. Consider previously-inactive and previously-active as the state of the slots
+// before the flipping happens.
+//
+// The flipping steps are the following:
+//
+// 1. Release the mutex for the previously-inactive profile slot. That slot, as seen by the sampler thread, is now
+// active.
+//
+// 2. Grab the mutex for the previously-active profile slot. Note that this can lead to the serializer thread blocking,
+// if the sampler thread is holding this mutex. After the mutex is grabbed, the previously-active slot becomes inactive,
+// as seen by the sampler thread.
+//
+// 3. Update `active_slot`.
+//
+// After flipping the profile slots, the serializer thread is now free to serialize the inactive profile slot. The slot
+// is kept inactive until the next time the serializer thread wants to serialize data.
+//
+// Note there can be a brief period between steps 1 and 2 where the serializer thread holds no lock, which means that
+// the sampler thread can pick either slot. This is OK: if the sampler thread picks the previously-inactive slot, the
+// samples will be reported on the next serialization; if the sampler thread picks the previously-active slot, the
+// samples are still included in the current serialization. Either option is correct.
+//
+// ### Additional notes
+//
+// Q: Can the sampler thread and the serializer thread ever be the same thread? (E.g. sampling in interrupt handler)
+// A: No; the current profiler design requires that sampling happens only on the thread that is holding the Global VM
+// Lock (GVL). The serializer thread flipping occurs after the serializer thread releases the GVL, and thus the
+// serializer thread will not be able to host the sampling process.
+//
+// ---
+
 static VALUE ok_symbol = Qnil; // :ok in Ruby
 static VALUE error_symbol = Qnil; // :error in Ruby
 
 static ID ruby_time_from_id; // id of :ruby_time_from in Ruby
 
 static VALUE stack_recorder_class = Qnil;
 
+// Contains native state for each instance
+struct stack_recorder_state {
```

Comment:
Why not use a single mutex to protect access to active profile/active_slot (unactive profile does not need protection since it is accessed only by serializing thread) ?

Something like:
```
struct stack_recorder_state {
  short active_slot;
  pthread_mutext_t mutex;
  ddprof_ffi_Profile* profiles[2];
};

ddprof_ffi_Profile* lock_active_profile(struct stack_recorder_state *state) {
  pthread_mutex_lock(state->mutex);
  return &profiles[active_slot];
}

void unlock_active_profile(struct stack_recorder_state *state) {
  pthread_mutex_unlock(state->mutex);
}

ddprof_ffi_Profile* flip_active_profile(struct stack_recorder_state *state) {
   lock_active_profile(state);
   active_slot = 1 - active_slot;
   unlock_active_profile(state);
   return &profiles[1 - active_slot];
}

void record_sample(VALUE recorder_instance, ddprof_ffi_Sample sample) {
  ...
  ddprof_ffi_Profile * profile = lock_active_profile(state);
  ddprof_ffi_Profile_add(profile, sample);
  unlock_active_profile(state);
}

static void *call_serialize_without_gvl(void *call_args) {
  struct call_serialize_without_gvl_arguments *args = (struct call_serialize_without_gvl_arguments *) call_args;
  args->profile = flip_active_profile(args->state);
  args->result = ddprof_ffi_Profile_serialize(args->profile, NULL /* end_time is optional */, NULL /* duration_nanos is optional */);
  args->serialize_ran = true;

  return NULL; // Unused
}
```

Reply:
This is an interesting alternative. I'd say the trade-off is that the solution in the PR will never cause the sampler thread to block, whereas this alternative might (if only for a very brief period).

E.g., considering a scheduler that has had a bad day, this can happen:

| sampler thread                                           | serializer thread                             |
|----------------------------------------------------------|-----------------------------------------------|
|                                                          | flip_active_profile                           |
|                                                          | lock_active_profile(state) -> mutex is locked |
|                                                          | **OS scheduler deschedules thread**           |
| record_sample                                            |                                               |
| lock_active_profile(state)                               |                                               |
| **thread is blocked until serializer thread gets to run again** |                                               |

I guess this could be solved by perhaps accepting the odd dropped sample -- e.g. if we used `trylock` for the sampler thread, and kept a regular lock for the serializer, we would never block, and the probability that the serializer thread would hold on to the lock for a long time is really small.

But the additional complexity in the current design I'd say comes from choosing never to block the serializer thread.

I'm slightly leaning towards the current solution for that reason, but I'm definitely not married to it (if it takes 1000 words to explain it, it's not simple, and I don't like that at all :P).

Reply:
I like the guarantee that the sampler thread can never block.
I believe that this is the most important requirement here: minimize impact on the host application's critical path.

Unless we start seeing bugs in the current implementation, I'm ok with the more complex design present here.

I do want to acknowledge that the solution by @nsavoire is much, much cleaner to read, but doesn't achieve the exact same result. I'm happy to revisit it if the current solution is too hard to maintain.

Reply:
Thanks @nsavoire for the feedback, I really mean it! Being critical about the design helps a lot in identifying gaps! 👍